File: tools/src/import/decode.ts
/**
 * Decode a ListForge share payload into a JSON object.
 *
 * ListForge packs a roster as `base64( gzip( utf8(json) ) )` and embeds it in a
 * URL hash fragment: `https://app/#/listforge/<BASE64>`. The fragment is used
 * deliberately so browsers never send it to a server, preserving the payload
 * verbatim. A valid gzipped payload always base64-encodes to a string starting
 * with `H4sIAAAAAAAAA`.
 *
 * {@link decodeListForge} accepts any of three forms and returns the parsed JSON:
 * - a full URL (the segment after the last `/` is taken),
 * - a bare base64 segment,
 * - an already-decoded JSON string (passed straight to `JSON.parse`).
 *
 * Decompression uses `fflate` (a tiny, dependency-free inflate) rather than
 * `node:zlib` so the importer works in browsers as well as Node — this module
 * is reachable from the package's root barrel and must stay universal.
 *
 * @packageDocumentation
 */
import { gunzipSync } from "fflate";

/**
 * Universal base64 → bytes. Node decodes via Buffer; browsers via atob.
 * (`Uint8Array.fromBase64` is still too new to rely on.)
 */
function base64ToBytes(b64: string): Uint8Array {
  if (typeof Buffer !== "undefined") {
    return new Uint8Array(Buffer.from(b64, "base64"));
  }
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i += 1) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

/** The base64 prefix every ListForge gzip payload begins with. */
const GZIP_BASE64_PREFIX = "H4sIA";

/** The path marker ListForge uses ahead of the payload. */
const LISTFORGE_MARKER = "/listforge/";

/**
 * Extract the payload segment from an input that may be a URL.
 *
 * The base64 alphabet includes `/`, so a bare base64 segment cannot be split on
 * `/`. We only treat the input as a URL when it carries the `/listforge/` marker
 * or an `http(s)://` scheme; otherwise it is returned unchanged.
 */
function extractSegment(input: string): string {
  const markerIndex = input.indexOf(LISTFORGE_MARKER);
  if (markerIndex !== -1) {
    return input.slice(markerIndex + LISTFORGE_MARKER.length);
  }
  if (/^https?:\/\//i.test(input)) {
    const lastSlash = input.lastIndexOf("/");
    return lastSlash === -1 ? input : input.slice(lastSlash + 1);
  }
  return input;
}

/**
 * Decode a ListForge payload (URL, bare base64, or raw JSON) into a JSON value.
 *
 * @throws if the input is neither valid JSON nor a decodable gzip payload.
 */
export function decodeListForge(input: string): unknown {
  const trimmed = input.trim();
  if (trimmed === "") {
    throw new Error("decodeListForge: empty input");
  }

  // Raw JSON object passed directly.
  if (trimmed.startsWith("{")) {
    return JSON.parse(trimmed);
  }

  const segment = extractSegment(trimmed);

  if (!segment.startsWith(GZIP_BASE64_PREFIX)) {
    throw new Error(
      "decodeListForge: input is not a ListForge payload (expected raw JSON, " +
        `or a gzip+base64 segment beginning with "${GZIP_BASE64_PREFIX}…")`,
    );
  }

  let json: string;
  try {
    const bytes = base64ToBytes(segment);
    json = new TextDecoder().decode(gunzipSync(bytes));
  } catch (cause) {
    throw new Error("decodeListForge: failed to gunzip base64 payload", {
      cause,
    });
  }

  return JSON.parse(json);
}

File: python/src/wh40kdc/imports/decode.py
"""Decode a ListForge share payload into a JSON object.

ListForge packs a roster as ``base64( gzip( utf8(json) ) )`` and embeds it in
a URL hash fragment: ``https://app/#/listforge/<BASE64>``. A valid gzipped
payload always base64-encodes to a string starting with ``H4sIA``.

:func:`decode_listforge` accepts any of three forms and returns the parsed
JSON: a full URL (the segment after the marker / last ``/`` is taken), a bare
base64 segment, or an already-decoded JSON string.

Python mirror of ``tools/src/import/decode.ts`` (stdlib ``base64``/``gzip``
only).
"""

from __future__ import annotations

import base64
import gzip
import json
import re
from typing import Any

#: The base64 prefix every ListForge gzip payload begins with.
GZIP_BASE64_PREFIX = "H4sIA"

#: The path marker ListForge uses ahead of the payload.
LISTFORGE_MARKER = "/listforge/"

_URL_RE = re.compile(r"^https?://", re.IGNORECASE)


def _extract_segment(input: str) -> str:
    """Extract the payload segment from an input that may be a URL.

    The base64 alphabet includes ``/``, so a bare base64 segment cannot be
    split on ``/``. The input is treated as a URL only when it carries the
    ``/listforge/`` marker or an ``http(s)://`` scheme.
    """
    marker_index = input.find(LISTFORGE_MARKER)
    if marker_index != -1:
        return input[marker_index + len(LISTFORGE_MARKER) :]
    if _URL_RE.match(input):
        last_slash = input.rfind("/")
        return input if last_slash == -1 else input[last_slash + 1 :]
    return input


def decode_listforge(input: str) -> Any:
    """Decode a ListForge payload (URL, bare base64, or raw JSON) into a JSON value.

    Raises ``ValueError`` if the input is neither valid JSON nor a decodable
    gzip payload.
    """
    trimmed = input.strip()
    if trimmed == "":
        raise ValueError("decode_listforge: empty input")

    # Raw JSON object passed directly.
    if trimmed.startswith("{"):
        return json.loads(trimmed)

    segment = _extract_segment(trimmed)

    if not segment.startswith(GZIP_BASE64_PREFIX):
        raise ValueError(
            "decode_listforge: input is not a ListForge payload (expected raw "
            "JSON, or a gzip+base64 segment beginning with "
            f'"{GZIP_BASE64_PREFIX}…")'
        )

    try:
        payload = gzip.decompress(base64.b64decode(segment))
    except Exception as cause:
        raise ValueError("decode_listforge: failed to gunzip base64 payload") from cause

    return json.loads(payload.decode("utf-8"))
